client/components/tinymce/plugins/mentions/mentions.jsx

```diff
@@ -0,0 +1,330 @@
+ */
+const VK = tinymce.util.VK;
+
+const getMatchingSuggestions = function( suggestions, query ) {
+	if ( ! query ) {
+		return suggestions.slice( 0, 10 );
+	}
+
+	return suggestions
+		.filter( ( { user_login, display_name } ) => {
+			// Start of string or preceded by a space.
+			const matcher = new RegExp( `^${ query }|\\s${ query }`, 'ig' );
+
+			return matcher.test( `${ user_login } ${ display_name }` );
+		} )
+		.slice( 0, 10 );
+};
+
+export class Mentions extends React.Component {
+	state = {
+		query: '',
+		popoverContext: null,
+		showPopover: false,
+	};
+
+	constructor( props ) {
+		super( props );
+	}
+
+	componentDidMount() {
+		const { editor } = this.props;
+		const { left, top } = this.getPosition();
+
+		editor.on( 'keyup', this.onKeyUp );
+		editor.on( 'click', this.hidePopover );
+
+		this.left = left;
+		this.top = top;
+	}
+
+	componentWillUpdate( nextProps, nextState ) {
+		// Update position of popover if going from invisible to visible state.
+		if ( ! this.state.showPopover && nextState.showPopover ) {
+			this.updatePosition( nextState );
+			this.props.editor.on( 'keydown', this.onKeyDown );
+
+			return;
+		}
+
+		// Visible to invisible state.
+		if ( this.state.showPopover && ! nextState.showPopover ) {
+			this.props.editor.off( 'keydown', this.onKeyDown );
+
+			return;
+		}
+
+		// Update position of popover if cursor has moved to a new line.
+		if ( nextState.showPopover ) {
+			const currentPosition = this.getPosition();
+			const isLineBefore = ( this.top > currentPosition.top ) && ( this.left < currentPosition.left );
+			const isLineAfter = ( this.top < currentPosition.top ) && ( this.left > currentPosition.left );
+
+			if ( isLineBefore || isLineAfter ) {
+				this.updatePosition( nextState, currentPosition );
+			}
+		}
+	}
+
+	componentWillUnmount() {
+		const { editor } = this.props;
+
+		editor.off( 'keydown', this.onKeyDown );
+		editor.off( 'keyup', this.onKeyUp );
+		editor.off( 'click', this.hidePopover );
+	}
+
+	getPosition( { query } = this.state ) {
+		const { editor } = this.props;
+		const range = editor.selection.getRng();
+		const mentionRange = document.createRange();
+
+		// Set range to start at beginning of mention in order to get accurate positioning values.
+		mentionRange.setStart( range.startContainer, range.startOffset - query.length );
+		mentionRange.setEnd( range.endContainer, range.endOffset );
+
+		const rectList = mentionRange.getClientRects();
+
+		if ( rectList.length > 0 ) {
+			const index = rectList.length - 1;
+
+			return {
+				left: rectList[ index ].left,
+				top: rectList[ index ].top,
+			};
+		}
+
+		const nodePosition = tinymce.$( editor.selection.getNode() ).offset();
+
+		return {
+			left: nodePosition.left,
+			top: nodePosition.top,
+		};
+	}
+
+	updatePosition( state, { left, top } = this.getPosition( state ) ) {
+		const { editor, node } = this.props;
+		const mceToolbar = tinymce.$( '.mce-toolbar-grp', editor.getContainer() )[ 0 ];
```

Comment:
It's unlikely to happen, but should we bulletproof this if we can't find the node?